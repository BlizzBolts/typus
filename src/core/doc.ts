import ts from "typescript";
import type { SourceFile } from "./sourceFile";

export interface SerialziedDoc {
  name?: string;
  documentation?: string;
  type?: string;
  tags?: { name: string; text: string }[];
  required?: boolean;
  members?: SerialziedDoc[];
  parameters?: SerialziedDoc[];
  children?: SerialziedDoc[];
  symbol?: ts.Symbol;
}

export interface DocPayload {
  name?: string;
  documentation?: string;
  type?: string;
  tags?: { name: string; text: string }[];
  required?: boolean;
  members?: Doc[];
  parameters?: Doc[];
  children?: Doc[];
  symbol?: ts.Symbol;
}

export class Doc {
  name?: string;
  documentation?: string;
  type?: string;
  tags?: { name: string; text: string }[];
  required?: boolean;
  members?: Doc[];
  parameters?: Doc[];
  children?: Doc[];
  symbol?: ts.Symbol;

  root?: SourceFile | null;

  constructor(param: DocPayload) {
    this.name = param.name || "";
    this.documentation = param.documentation || "";
    this.type = param.type || "";
    this.tags = param.tags || [];
    this.required = param.required || false;
    this.members = param.members || [];
    this.parameters = param.parameters || [];
    this.children = param.children || [];
    this.symbol = param.symbol || (null as unknown as ts.Symbol);

    this.root = null;
  }

  public setRoot(root: SourceFile) {
    this.root = root;
  }

  static serialize = (curr: Doc): SerialziedDoc => {
    return {
      name: curr.name,
      documentation: curr.documentation,
      type: curr.type,
      tags: curr.tags,
      required: curr.required,
      members: curr.members?.map(Doc.serialize),
      parameters: curr.parameters?.map(Doc.serialize),
      children: curr.children?.map(Doc.serialize),
    };
  };
}
