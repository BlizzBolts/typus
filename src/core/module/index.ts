export { ModuleGraph } from "./graph";
export { Module } from "./module";
export type { BaseModule } from "./types";
