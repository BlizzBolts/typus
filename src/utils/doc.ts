import ts from "typescript"
import type { DocEntry } from "../types"

// export const getInterfaceDocuments = (
//   node: ts.Node,
//   checker: ts.TypeChecker,
// ) => {
//   const symbol = checker.getSymbolAtLocation(node.name)
//   if (!symbol) {
//     return
//   }

//   const main = serializeSymbol(symbol, checker, node)

//   main.fields = node.members
//     .map((m) => {
//       if (!m.name) return
//       const symbol = checker.getSymbolAtLocation(m.name)
//       if (!symbol) return
//       return serializeSymbol(symbol, checker, m)
//     })
//     .filter(Boolean) as DocEntry[]

//   return main
// }

export const serializeSymbol = (
  symbol: ts.Symbol,
  checker: ts.TypeChecker,
): DocEntry => {
  return {
    name: symbol?.getName(),
    documentation: ts.displayPartsToString(
      symbol?.getDocumentationComment(checker),
    ),
    symbolType: checker.getTypeOfSymbolAtLocation(
      symbol,
      symbol?.valueDeclaration!,
    ),
    type: checker.typeToString(
      checker.getTypeOfSymbolAtLocation(symbol, symbol?.valueDeclaration!),
    ),
    tags: symbol?.getJsDocTags(checker).map((o) => {
      return {
        name: o.name,
        text: ts.displayPartsToString(o.text),
      }
    }),
    children: [],
    members: [],
    parameters: [],
    // @ts-ignore
    required: symbol?.["questionToken"],
  }
}
