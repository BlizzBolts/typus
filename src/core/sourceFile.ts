import { Doc, SerialziedDoc } from "./doc";

export type FilePath = string;

interface SourceFilePayload {
  filePath: FilePath;
  docs?: Doc[];
}

export interface SerialzieSourceFile {
  filePath: FilePath;
  docs: SerialziedDoc[];
}

export class SourceFile {
  private filePath: FilePath;

  private docs?: Doc[];

  constructor(payload: SourceFilePayload) {
    const { filePath, docs = [] } = payload;
    this.filePath = filePath;
    this.docs = docs;
  }

  public setFilePath(filePath: FilePath): this {
    this.filePath = filePath;
    return this;
  }

  public getFilePath(): FilePath {
    return this.filePath;
  }

  public addDoc(doc: Doc) {
    this.docs!.push(doc);
    return this;
  }

  public static serialize(curr: SourceFile) {
    return {
      filePath: curr.filePath,
      docs: curr.docs?.map(Doc.serialize),
    };
  }
}
