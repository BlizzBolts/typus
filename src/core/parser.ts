import ts from "typescript";
import { addCache, getCache, hasCache } from "../cache";
import { serializeSymbol } from "../utils/doc";
import { Doc } from "./doc";

export const defaultCompilerOptions = {
  target: ts.ScriptTarget.ES5,
  module: ts.ModuleKind.CommonJS,
};

export class Parser {
  private compilerOptions: ts.CompilerOptions;
  private filePaths: string | string[];
  private program: ts.Program;
  private typeChecker: ts.TypeChecker;

  constructor(
    filePaths: string | string[],
    compilerOptions: Partial<ts.CompilerOptions>
  ) {
    this.compilerOptions = Object.assign(
      defaultCompilerOptions,
      compilerOptions
    );
    this.filePaths = Array.isArray(filePaths) ? filePaths : [filePaths];
    this.program = ts.createProgram(this.filePaths, this.compilerOptions);
    this.typeChecker = this.program.getTypeChecker();
  }

  parse() {
    const validSourceFiles = this.program.getSourceFiles().filter((o) => {
      return !o.isDeclarationFile;
    });

    const rootDoc = new Doc();

    validSourceFiles.forEach((o) => {
      o.forEachChild((p) => {
        this.traverse(p, rootDoc);
      });
    });

    return rootDoc;
  }

  parseInterface(node: ts.InterfaceDeclaration, parent: Doc) {
    const interfaceType = this.typeChecker.getTypeAtLocation(node.name);
    const Doc = serializeSymbol(interfaceType.getSymbol()!, this.typeChecker);
    parent.children?.push(Doc);
    addCache(interfaceType, Doc);

    if (Array.isArray(node.members)) {
      node.members.forEach((memberNode) => {
        this.traverse(memberNode, Doc);
      });
    }
  }

  parseFunction(node: ts.FunctionDeclaration, parent: Doc) {
    const fnType = this.typeChecker.getTypeAtLocation(node);
    const Doc = serializeSymbol(fnType.getSymbol()!, this.typeChecker);
    parent.children?.push(Doc);
    addCache(fnType, Doc);

    if (Array.isArray(node.parameters)) {
      node.parameters.forEach((parameter) => this.traverse(parameter, Doc));
    }
  }

  parseFunctionParameter = (node: ts.ParameterDeclaration, parent: Doc) => {
    const parameterType = this.typeChecker.getTypeAtLocation(node.name);
    if (hasCache(parameterType)) {
      parent.parameters?.push(getCache(parameterType)!);
    } else {
      const symbol = this.typeChecker.getSymbolAtLocation(node.name!);
      parent.parameters?.push(serializeSymbol(symbol!, this.typeChecker));
    }
  };

  parseInterfaceMember = (node: ts.TypeElement, parent: Doc) => {
    const memberType = this.typeChecker.getTypeAtLocation(node!.name!);

    if (hasCache(memberType)) {
      parent.members?.push(getCache(memberType)!);
    } else {
      const symbol = this.typeChecker.getSymbolAtLocation(node.name!);
      parent.members?.push(serializeSymbol(symbol!, this.typeChecker));
    }
  };

  traverse(node: ts.Node, parent: Doc) {
    if (ts.isInterfaceDeclaration(node)) {
      this.parseInterface(node, parent);
    }

    if (ts.isFunctionDeclaration(node)) {
      this.parseFunction(node, parent);
    }

    if (ts.isParameter(node)) {
      this.parseFunctionParameter(node, parent);
    }

    if (ts.isTypeElement(node)) {
      this.parseInterfaceMember(node, parent);
    }

    if (ts.isVariableDeclaration(node)) {
      console.log("variable!");
    }
  }

  getProgram() {
    return this.program;
  }

  getTypeChecker() {
    return this.typeChecker;
  }
}
