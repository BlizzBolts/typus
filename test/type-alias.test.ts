import { expect, it, suite } from "vitest";
import path from "path";
import { Parser } from "../src/core/parser";
import { Doc } from "../src/core/doc";

suite("type alias", () => {
  it("should parse type alias string", () => {
    const parser = Parser.of().setup(
      path.resolve(__dirname, "./manifest/type-alias/string.ts")
    );
    const docCache = parser.parse();
    expect(Doc.serialize(docCache)).toMatchSnapshot();
  });
  it("should parse type alias number", () => {});
  it("should parse type alias boolean", () => {});
  it("should parse type alias any", () => {});
  it("should parse type alias unknown", () => {});
  it("should parse type alias undefined", () => {});
  it("should parse type alias null", () => {});
  it("should parse type alias a function", () => {});
  it("should parse type alias array of string", () => {});
});
