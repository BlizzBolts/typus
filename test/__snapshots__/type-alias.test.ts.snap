// Vitest Snapshot v1

exports[`type alias > should parse type alias string 1`] = `
{
  "children": [],
  "documentation": "",
  "members": [],
  "name": "root",
  "parameters": [],
  "required": false,
  "tags": [],
  "type": "",
}
`;
